### PR TITLE
Added start time for supporting restarts for fixed rate schedules

### PR DESCRIPTION
**Pull request by pmahindrakar-oss**
Signed-off-by: Prafulla Mahindrakar <prafulla.mahindrakar@gmail.com>

# TL;DR
_Please replace this text with a description of what this PR accomplishes._

## Type
 - [X] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [X] Smoke tested
 - [X] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes [https://github.com/flyteorg/flyte/issues/<number>](https://github.com/flyteorg/flyte/issues/2885)

### CODE REVIEW

**Comment by katrogan** [comment]
can you add a tracking issue?

File: scheduler/core/gocron_scheduler_test.go [hunk 1]
```diff
@@ -276,3 +284,74 @@ func TestCatchUpAllSchedule(t *testing.T) {
...
+			wg.Done()
+			return nil
+		}
+		c := cron.New()
+		g := GoCronScheduler{cron: c, jobStore: sync.Map{}}
+		err := g.ScheduleJob(ctx, scheduleFixed, timedFuncWithSchedule, &lastTime)
+		c.Start()
+		assert.NoError(t, err)
+		select {
+		case <-time.After(time.Minute * 2):
```
Comment (katrogan) [inline scheduler/core/gocron_scheduler_test.go:320]: does this have to be such a long interval for unit tests?

Comment (pmahindrakar-oss) [reply scheduler/core/gocron_scheduler_test.go:320]: This will 2 min only in failure case.This is more of an integration test since there is no easy way to unit test this particular behaviour

Comment (pmahindrakar-oss) [reply scheduler/core/gocron_scheduler_test.go]: Move this to integration test package and wont impact the unit tests

File: scheduler/core/gocron_scheduler.go [hunk 2]
```diff
@@ -271,7 +271,11 @@ func (g *GoCronScheduler) AddFixedIntervalJob(ctx context.Context, job *GoCronJo
 	var jobFunc cron.TimedFuncJob
 	jobFunc = job.Run
 
-	entryID := g.cron.ScheduleTimedJob(cron.ConstantDelaySchedule{Delay: d}, jobFunc)
+	var lastExecTime time.Time
+	if job.lastTime != nil {
```
Comment (katrogan) [inline scheduler/core/gocron_scheduler.go:275]: nit: should this field name change to job.lastExecTime too?

File: tests/scheduler_test.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+		assert.NoError(t, err)
+		select {
+		case <-time.After(time.Minute * 2):
+			assert.Fail(t, "timed job didn't get triggered")
+		case <-wait(wg):
+			break
+		}
+	})
+
+	t.Run("without schedule time", func(t *testing.T) {
```
Comment (katrogan) [inline tests/scheduler_test.go:73]: sorry if I'm being dense - can you point out how this test differs than the one above?

**Review by katrogan — DISMISSED** [review]

File: tests/scheduler_test.go (hunk 3 above)
Comment (pmahindrakar-oss) [reply tests/scheduler_test.go:73]: check L62 & L90, 
The last Parameter to schedulerJob is different.

File: scheduler/core/gocron_scheduler.go (hunk 2 above)
Comment (pmahindrakar-oss) [reply scheduler/core/gocron_scheduler.go:275]: Changing this would mean maintaining backward compat. I will clean this up when i next make change to this file

Comment (pmahindrakar-oss) [reply scheduler/core/gocron_scheduler.go:275]: On second thought, the JobStore will get loaded again after new version, so it will start using the new field after the restart, so we don't have to worry about backward compat.

File: tests/scheduler_test.go (hunk 3 above)
Comment (pmahindrakar-oss) [reply tests/scheduler_test.go]: Refactored the test to remove redundant code which makes this test clearer

**Review by katrogan — DISMISSED** [review]

File: scheduler/core/gocron_scheduler.go (hunk 2 above)
Comment (katrogan) [reply scheduler/core/gocron_scheduler.go]: Awesome, can you verify that there are indeed no problems on restart locally?

**Review by eapolinario — APPROVED** [review]
